Mathlib/Algebra/Hom/Group.lean

```diff
@@ -0,0 +1,1712 @@
+  Function.Injective (MonoidWithZeroHom.toZeroHom : (M →*₀ N) → ZeroHom M N) :=
+  fun _ _ h => MonoidWithZeroHom.ext <| ZeroHom.ext_iff.mp h
+#align monoid_with_zero_hom.to_zero_hom_injective MonoidWithZeroHom.to_zero_hom_injective
+
+@[simp, to_additive]
+theorem OneHom.comp_id [One M] [One N] (f : OneHom M N) : f.comp (OneHom.id M) = f :=
+  OneHom.ext fun _ => rfl
+#align one_hom.comp_id OneHom.comp_id
+
+@[simp, to_additive]
+theorem MulHom.comp_id [Mul M] [Mul N] (f : M →ₙ* N) : f.comp (MulHom.id M) = f :=
+  MulHom.ext fun _ => rfl
+#align mul_hom.comp_id MulHom.comp_id
+
+@[simp, to_additive]
+theorem MonoidHom.comp_id [MulOneClass M] [MulOneClass N] (f : M →* N) :
+  f.comp (MonoidHom.id M) = f := MonoidHom.ext fun _ => rfl
+#align monoid_hom.comp_id MonoidHom.comp_id
+
+@[simp]
+theorem MonoidWithZeroHom.comp_id [MulZeroOneClass M] [MulZeroOneClass N] (f : M →*₀ N) :
+  f.comp (MonoidWithZeroHom.id M) = f := MonoidWithZeroHom.ext fun _ => rfl
+#align monoid_with_zero_hom.comp_id MonoidWithZeroHom.comp_id
+
+@[simp, to_additive]
+theorem OneHom.id_comp [One M] [One N] (f : OneHom M N) : (OneHom.id N).comp f = f :=
+  OneHom.ext fun _ => rfl
+#align one_hom.id_comp OneHom.id_comp
+
+@[simp, to_additive]
+theorem MulHom.id_comp [Mul M] [Mul N] (f : M →ₙ* N) : (MulHom.id N).comp f = f :=
+  MulHom.ext fun _ => rfl
+#align mul_hom.id_comp MulHom.id_comp
+
+@[simp, to_additive]
+theorem MonoidHom.id_comp [MulOneClass M] [MulOneClass N] (f : M →* N) :
+  (MonoidHom.id N).comp f = f := MonoidHom.ext fun _ => rfl
+#align monoid_hom.id_comp MonoidHom.id_comp
+
+@[simp]
+theorem MonoidWithZeroHom.id_comp [MulZeroOneClass M] [MulZeroOneClass N] (f : M →*₀ N) :
+  (MonoidWithZeroHom.id N).comp f = f := MonoidWithZeroHom.ext fun _ => rfl
+#align monoid_with_zero_hom.id_comp MonoidWithZeroHom.id_comp
+
+-- Porting note: restore `to_additive`
+-- @[to_additive AddMonoidHom.map_nsmul]
+protected theorem MonoidHom.map_pow [Monoid M] [Monoid N] (f : M →* N) (a : M) (n : ℕ) :
+  f (a ^ n) = f a ^ n := map_pow f a n
+#align monoid_hom.map_pow MonoidHom.map_pow
+protected theorem AddMonoidHom.map_nsmul [AddMonoid M] [AddMonoid N] (f : M →+ N) (a : M) (n : ℕ) :
+  f (n • a) = n • f a := map_nsmul f n a
+
+-- Porting note: restore `to_additive`
+-- @[to_additive]
+protected theorem MonoidHom.map_zpow' [DivInvMonoid M] [DivInvMonoid N] (f : M →* N)
+  (hf : ∀ x, f x⁻¹ = (f x)⁻¹) (a : M) (n : ℤ) :
+  f (a ^ n) = f a ^ n := map_zpow' f hf a n
+#align monoid_hom.map_zpow' MonoidHom.map_zpow'
+protected theorem AddMonoidHom.map_zsmul' [SubNegMonoid M] [SubNegMonoid N] (f : M →+ N)
+  (hf : ∀ x, f (-x) = -(f x)) (a : M) (n : ℤ) :
+  f (n • a) = n • f a := map_zsmul' f hf a n
+
+section End
+
+-- Porting note: this part causes instance loop
+-- namespace Monoid
+
+-- variable (M) [MulOneClass M]
+
+-- /-- The monoid of endomorphisms. -/
+-- protected def End := M →* M
+-- #align monoid.End Monoid.End
+
+-- namespace End
+
+-- instance : Monoid (Monoid.End M) where
```

Comment:
This instance causes the error
```
(deterministic) timeout at 'whnf', maximum number of heartbeats (200000) has been reached (use 'set_option maxHeartbeats <num>' to set the limit)
```
in `AddMonoid.coe_mul`, `MonoidHom.mul_comp`, and `MonoidHom.comp_mul` later.